includes/enrolment/class-sensei-enrolment-provider-journal-store.php

```diff
@@ -0,0 +1,283 @@
+<?php
+/**
+ * File containing the class Sensei_Enrolment_Provider_Journal_Store.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * This class is responsible for storing provider metadata like logs and status history.
+ */
+class Sensei_Enrolment_Provider_Journal_Store implements JsonSerializable {
+	const META_PREFIX_ENROLMENT_PROVIDERS_JOURNAL = 'sensei_enrolment_providers_journal_';
+
+	/**
+	 * Flag for if a state the store has changed.
+	 *
+	 * @var bool
+	 */
+	private $has_changed = false;
+
+	/**
+	 * Journal objects for the providers.
+	 *
+	 * @var Sensei_Enrolment_Provider_Journal[]
+	 */
+	private $providers_journal;
+
+	/**
+	 * User ID that this store is used for.
+	 *
+	 * @var int
+	 */
+	private $user_id;
+
+	/**
+	 * Course post ID that this store is used for.
+	 *
+	 * @var int
+	 */
+	private $course_id;
+
+	/**
+	 * Keeps track of instances of this class.
+	 *
+	 * @var self[][]
+	 */
+	private static $instances = [];
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function __construct( $user_id, $course_id ) {
+		$this->user_id           = $user_id;
+		$this->course_id         = $course_id;
+		$this->providers_journal = [];
+	}
+
+	/**
+	 * Get a journal store record for a user/course.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return self
+	 */
+	private static function get( $user_id, $course_id ) {
```

Comment:
Differing from the state store, it looks like the individual provider states are generally inaccessible, is that right? The only way a provider can interact with its journal is via the public static methods in this class. Was that an intentional design decision? 

For the state store, providers can access their own state from `\Sensei_Course_Enrolment::get_provider_state`.

Reply:
Yep. The reasons that I chose this approach were that I think that it provides a bit better encapsulation since it doesn't expose `Sensei_Enrolment_Provider_Journal`'s API and that it will be easier to implement cross course/user functionality. It also doesn't require from callers to call `get` first.